concent_api/common/decorators.py

```diff
@@ -54,13 +54,38 @@ def wrapper(*args: Any, **kwargs: Any) -> None:
     return wrapper
 
 
-def non_nesting_atomic(using: Union[str, Callable], savepoint: bool=True) -> Atomic:
-    if (
-        settings.DETECT_NESTED_TRANSACTIONS and
-        get_connection(using).in_atomic_block
-    ):
-        raise ConcentPendingTransactionError
-    if callable(using):
-        return Atomic(DEFAULT_DB_ALIAS, savepoint)(using)
-    else:
-        return Atomic(using, savepoint)
+class non_nesting_atomic(ContextDecorator):
+    """
+    This class acts as either decorator or context manager.
+    It discovers when the callable or code is being run in nested database transaction and raises exception.
+    Otherwise it wraps callable or code in database transaction.
+    """
+
+    def __init__(self, using: Union[str, Callable], savepoint: bool=True):
+        self.using = using
+        self.savepoint = savepoint
+        self.atomic_context_decorator = Atomic(self.using, self.savepoint)
+
+    def __enter__(self) -> None:
+        if (
+            settings.DETECT_NESTED_TRANSACTIONS and
+            get_connection(self.using).in_atomic_block
+        ):
+            raise ConcentPendingTransactionError
+        return self.atomic_context_decorator.__enter__()
+
+    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+        return self.atomic_context_decorator.__exit__(*args, **kwargs)
+
+    def __call__(self, func: Callable) -> Callable:
+        @wraps(func)
+        def decorated(*args: Any, **kwargs: Any):
+            if (
```

Comment:
this `if` is redundant and unnecessary... please remove it.

Reply:
Why? It is called when you use this class as decorator.

Reply:
:+1: good catch